besskge/pipeline.py

```diff
@@ -0,0 +1,279 @@
+
+            batch_res = []
+            batch_idx = []
+            for i in range(self.bess_module.n_step):
+                step = (
+                    torch.tensor([i], dtype=torch.int32)
+                    .broadcast_to(
+                        (
+                            self.bess_module.sharding.n_shard
+                            * self.batch_sampler.batches_per_step,
+                            1,
+                        )
+                    )
+                    .contiguous()
+                )
+                ent_slice = torch.minimum(
+                    i * self.bess_module.window_size
+                    + torch.arange(self.bess_module.window_size),
+                    torch.tensor(self.bess_module.sharding.max_entity_per_shard - 1),
+                )
+                # Global indices of entities scored in the step
+                batch_idx.append(
+                    self.bess_module.sharding.shard_and_idx_to_entity[
+                        :, ent_slice
+                    ].flatten()
+                )
+                inp = {k: v.flatten(end_dim=1) for k, v in batch.items()}
+                inp.update(dict(step=step))
+                batch_res.append(self.poptorch_module(**inp))
+            batch_scores = torch.concat(batch_res, dim=-1)
+            # Filter out padding scores
+            batch_scores_filt = batch_scores[triple_mask.flatten()][
+                :, np.unique(np.concatenate(batch_idx), return_index=True)[1]
+            ][:, : self.bess_module.sharding.n_entity]
+            # Scores of positive triples
+            true_scores = batch_scores_filt[
+                torch.arange(batch_scores_filt.shape[0]),
+                ground_truth[triple_mask],
+            ]
+            if self.filter_triples is not None:
+                # Filter for triples in batch
+                batch_filter = get_entity_filter(
+                    self.triples[triple_id[triple_mask]],
+                    self.filter_triples,
+                    filter_mode=self.corruption_scheme,
+                )
+                batch_scores_filt[batch_filter[:, 0], batch_filter[:, 1]] = -torch.inf
+
+            if self.evaluation:
+                assert (
+                    ground_truth is not None
+                ), "Evaluation requires providing ground truth entities"
+                # If not already masked, mask scores of true triples
+                # to compute metrics
+                batch_scores_filt[
+                    torch.arange(batch_scores_filt.shape[0]),
+                    ground_truth[triple_mask],
+                ] = -torch.inf
+                batch_ranks = self.evaluation.ranks_from_scores(
+                    true_scores, batch_scores_filt
+                )
+                metrics.append(self.evaluation.dict_metrics_from_ranks(batch_ranks))
+                if self.evaluation.return_ranks:
+                    ranks.append(batch_ranks)
+            if self.return_scores:
```

Comment:
I think it would still be good to have the option of returning the topK predicted entities per query (after filtering). E.g. if we don't want to keep all scores for memory reasons but would like to check how many of the top 100 predictions are of the expected entity type.

Reply:
Good point, I thought that if one was interested in the topK, they could just run the more efficient `TopKQueryBess`, but that would not support filtering. I will add this option also here.

Reply:
Implemented `return_topk` option in [247d88b](https://github.com/graphcore-research/bess-kge/pull/33/commits/247d88b4c3a38d816eb66f409e6887d111671d5f)